# Mobile App Data Analysis
Our company builds Android and iOS mobile apps and the focus is on finding mobile app profiles that are profitable for App Store and Google Play market. We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that the number of users of our apps determines our revenue for any given app — the more users who see and engage with the ads, the better.

Our goal for this project is to analyze data to help our developers understand what type of apps are likely to attract more users.

In [1]:
from csv import reader

### Google Play dataset ###
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

### Apple Store dataset ###
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]


In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))


In [3]:
print(android_header,'\n')
explore_data(android,0,5,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Eve

Google Play dataset has 10841 apps and 13 columns. The columns that might be useful for our analysis are:
- App
- Category
- Reviews
- Installs
- Type
- Price
- Genres

Documentation regarding the data can be found [here](https://www.kaggle.com/datasets/lava18/google-play-store-apps)

In [4]:
print(ios_header,'\n')
explore_data(ios,0,5,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16


App Store dataset has 7197 apps and 16 columns. The columns that can help with our analysis are:
- track_name
- currency
- price
- rating_count_tot
- rating_count_ver
- prime_genre

Documentation regarding the data can be found [here](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps)

# Deleting wrong data
We are going to check for row which have incorrect number of columns and delete those. We will identify such rows by comparing the length of the row with the length of the header.

In [5]:
len_header = len(android_header)
print(len(android))
for idx,row in enumerate(android):
    if len(row) != len_header:
        print(f'Row is {row}, \nIndex is {idx}')
        del android[idx]
print(f'Length of android dataset after deleting wrong data: {len(android)}')

10841
Row is ['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up'], 
Index is 10472
Length of android dataset after deleting wrong data: 10840


Now lets do the same for ios data

In [6]:
len_header = len(ios_header)
print(len(ios))
for idx,row in enumerate(ios):
    if len(row) != len_header:
        print(f'Row is {row}, \nIndex is {idx}')
        del ios[idx]
print(f'Length of ios dataset after deleting wrong data: {len(ios)}')

7197
Length of ios dataset after deleting wrong data: 7197


# Identifying Duplicate Rows
We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app. In order to do that, we need to first check if we have any duplicate rows.


In [7]:
app_names = []
duplicates = []
for row in android:
    app_name = row[0]
    if app_name not in app_names:
        app_names.append(app_name)
    else:
        duplicates.append(app_name)
print(duplicates,len(duplicates))


['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express', 'Accounting App - Zoho Books', 'Invoice & Time Tracking - Zoho', 'join.me - Simple Meetings', 'Invoice 2go — Professional Invoices and Estimates', 'SignEasy | Sign and Fill PDF and other Documents', 'Quick PDF Scanner + OCR FREE', 'Genius Scan - PDF Scanner', 'Tiny Scanner - PDF Scanner App', 'Fast Scanner : Free PDF Scan', 'Mobile Doc Scanner (MDScan) Lite', 'TurboScan: scan documents and receipts in PDF', 'Tiny Scanner Pro: PDF Doc Scan', 'Docs To Go™ Free Office Suite', 'OfficeSuite : Free Office + PDF Editor',

Google play dataset has 1181 duplicate rows. One thing we could do is remove the duplicate rows randomly, but we could probably find a better way. Let's look at 'Instagram' app as an example to see what could duplicate rows mean.

In [8]:
for row in android:
    app_name = row[0]
    if app_name == 'Instagram':
        print(row,'\n')

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device'] 

['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device'] 

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device'] 

['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device'] 



Based on the rows we printed for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show the data was collected at different times. 

We can use this information to build a criterion for removing the duplicates. The higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, **we'll only keep the row with the highest number of reviews** and remove the other entries for any given app.

Expected number of rows after removing duplicates = `len(android) - 1181` = `9659`

# Removing Duplicate Rows
To do that, we will:

- Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
- Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)
- We need to add a supplementary condition (already_added list) to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [9]:
reviews_max = {}
for row in android:
    app = row[0]
    reviews = float(row[3])
    if app not in reviews_max :
        reviews_max[app] = reviews
    elif app in reviews_max and reviews_max[app] < reviews:
        reviews_max[app] = reviews
print(len(reviews_max))

9659


In [10]:
already_added=[]
android_clean = []
for row in android:
    app = row[0]
    reviews = float(row[3])
    if (reviews == reviews_max[app]) and (app not in already_added):
        android_clean.append(row)
        already_added.append(app)
print(len(android_clean))

9659


In [11]:
explore_data(android_clean, 0,3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


**Checking for duplicates in App Store dataset.**

In [12]:
ids = []
duplicates = []
for row in ios:
    id = row[0]
    if id not in ids:
        ids.append(id)
    else:
        duplicates.append(id)
print(duplicates,'\n',len(duplicates))

[] 
 0


No duplicates found in ios.

# Removing non English apps
Remember we use English for the apps we develop at our company, and we'd like to analyze only the apps that are designed for an English-speaking audience. So let's first find out if we have any apps names that suggest they are not designed for an English-speaking audience.

The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system. Based on this number range, we can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

If an app name contains a character that is greater than 127, then it probably means that the app has a non-English name. 

In [13]:
def is_english(s):
    for char in s:
        if ord(char) > 127:
            return False
    return True
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instagram'))
print(is_english('Instachat 😜'))

False
False
True
False


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.To minimize the impact of data loss, we'll only remove an app if its name has more than three non-ASCII characters. 



In [14]:
def is_english(s):
    count = 0
    for char in s:
        if ord(char) > 127:
            count += 1
    if count > 3:
        return False
    return True
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instagram'))
print(is_english('Instachat 😜'))

False
True
True
True


The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in our analysis — we shouldn't spend too much time on optimization at this point.

Below we use this filter on both the datasets.

In [15]:
android_english = []
ios_english = []
for row in android_clean:
    app = row[0]
    if is_english(app):
        android_english.append(row)
for row in ios:
    app = row[1]
    if is_english(app):
        ios_english.append(row)
explore_data(android_english, 0, 3, True)
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'G

We can see that we are now left with 9614 Android apps and 6183 iOS apps

# Isolating Free Apps
As we mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our datasets contain both free and non-free apps; we'll need to isolate only the free apps for our analysis.

In [16]:
android_final = []
ios_final = []
for row in android_english:
    price = row[7]
    if price == '0':
        android_final.append(row)
for row in ios_english:
    price = row[4]
    if price == '0.0':
        ios_final.append(row)
print(len(android_final))
print(len(ios_final))

8864
3222


We are now left with 8864 android apps and 3222 ios apps. This should be sufficient for our analysis.

# Most Common Apps By Genre
As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we then develop it further.
3. If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the `prime_genre` column of the App Store data set, and the `Genres` and `Category` columns of the Google Play data set.

We'll build two functions we can use to analyze the frequency tables:

1. One function to generate frequency tables that show percentages
2. Another function we can use to display the percentages in a descending order

In [17]:
def freq_table(data, idx):
    d = {}
    for row in data:
        val = row[idx]
        if val in d:
            d[val] += 1
        else:
            d[val] = 1
    for k in d:
        d[k] = d[k]*100/len(data)
    return d

def display_table(data, idx):
    table = freq_table(data, idx)
    table_sorted = {k:v for k,v in sorted(table.items(), key = lambda x:x[1], reverse = True)}
    for k,v in table_sorted.items():
        print(f'{k}:{v}')


In [18]:
display_table(ios_final,-5 )

Games:58.16263190564867
Entertainment:7.883302296710118
Photo & Video:4.9658597144630665
Education:3.6623215394165114
Social Networking:3.2898820608317814
Shopping:2.60707635009311
Utilities:2.5139664804469275
Sports:2.1415270018621975
Music:2.0484171322160147
Health & Fitness:2.017380509000621
Productivity:1.7380509000620732
Lifestyle:1.5828677839851024
News:1.3345747982619491
Travel:1.2414649286157666
Finance:1.1173184357541899
Weather:0.8690254500310366
Food & Drink:0.8069522036002483
Reference:0.5586592178770949
Business:0.5276225946617008
Book:0.4345127250155183
Navigation:0.186219739292365
Medical:0.186219739292365
Catalogs:0.12414649286157665


In [19]:
display_table(android_final, -4) # Genres

Tools:8.449909747292418
Entertainment:6.069494584837545
Education:5.347472924187725
Business:4.591606498194946
Lifestyle:3.892148014440433
Productivity:3.892148014440433
Finance:3.700361010830325
Medical:3.5311371841155235
Sports:3.463447653429603
Personalization:3.3167870036101084
Communication:3.237815884476534
Action:3.1024368231046933
Health & Fitness:3.079873646209386
Photography:2.9444945848375452
News & Magazines:2.7978339350180503
Social:2.6624548736462095
Travel & Local:2.3240072202166067
Shopping:2.2450361010830324
Books & Reference:2.1435018050541514
Simulation:2.041967509025271
Dating:1.861462093862816
Arcade:1.8501805054151625
Video Players & Editors:1.7712093862815885
Casual:1.759927797833935
Maps & Navigation:1.3989169675090252
Food & Drink:1.2409747292418774
Puzzle:1.128158844765343
Racing:0.9927797833935018
Libraries & Demo:0.9363718411552346
Role Playing:0.9363718411552346
Auto & Vehicles:0.9250902527075813
Strategy:0.9138086642599278
House & Home:0.8235559566787004
W

In [20]:
display_table(android_final, 1) # Category

FAMILY:18.907942238267147
GAME:9.724729241877256
TOOLS:8.461191335740072
BUSINESS:4.591606498194946
LIFESTYLE:3.9034296028880866
PRODUCTIVITY:3.892148014440433
FINANCE:3.700361010830325
MEDICAL:3.5311371841155235
SPORTS:3.395758122743682
PERSONALIZATION:3.3167870036101084
COMMUNICATION:3.237815884476534
HEALTH_AND_FITNESS:3.079873646209386
PHOTOGRAPHY:2.9444945848375452
NEWS_AND_MAGAZINES:2.7978339350180503
SOCIAL:2.6624548736462095
TRAVEL_AND_LOCAL:2.33528880866426
SHOPPING:2.2450361010830324
BOOKS_AND_REFERENCE:2.1435018050541514
DATING:1.861462093862816
VIDEO_PLAYERS:1.7937725631768953
MAPS_AND_NAVIGATION:1.3989169675090252
FOOD_AND_DRINK:1.2409747292418774
EDUCATION:1.1620036101083033
ENTERTAINMENT:0.9589350180505415
LIBRARIES_AND_DEMO:0.9363718411552346
AUTO_AND_VEHICLES:0.9250902527075813
HOUSE_AND_HOME:0.8235559566787004
WEATHER:0.8009927797833934
EVENTS:0.7107400722021661
PARENTING:0.6543321299638989
ART_AND_DESIGN:0.6430505415162455
COMICS:0.6204873646209387
BEAUTY:0.597924187

The difference between the `Genres` and the `Category` columns is not crystal clear, but one thing we can notice is that the `Genres` column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the `Category` column moving forward.

# Most Popular Apps
## On App Store
Now we'd like to get an idea about the kind of apps that have most users. One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the `Installs` column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the `rating_count_tot` app.

Below, we calculate the average number of user ratings per app genre on the App Store:

In [21]:
genres_ios = freq_table(ios_final, -5)
genres_rating = {}
for genre in genres_ios:
    total_rating = 0
    app_count = 0
    for row in ios_final:
        row_genre = row[-5]
        n_rating = float(row[5])
        if row_genre == genre:
            total_rating += n_rating
            app_count += 1
    avg_rating = total_rating/app_count
    genres_rating[genre] = avg_rating
genres_sorted = {k:v for k,v in sorted(genres_rating.items(), key = lambda x:x[1], reverse = True)}
for k,v in genres_sorted.items():
    print(f'{k} : {v}')
    

Navigation : 86090.33333333333
Reference : 74942.11111111111
Social Networking : 71548.34905660378
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 39758.5
Food & Drink : 33333.92307692308
Finance : 31467.944444444445
Photo & Video : 28441.54375
Travel : 28243.8
Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Sports : 23008.898550724636
Games : 22788.6696905016
News : 21248.023255813954
Productivity : 21028.410714285714
Utilities : 18684.456790123455
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Business : 7491.117647058823
Education : 7003.983050847458
Catalogs : 4004.0
Medical : 612.0


In [22]:
for row in ios_final:
    if row[-5] == 'Navigation':
        print(row[1], ':', row[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together.

In [23]:
for row in ios_final:
    if row[-5] == 'Social Networking':
        print(row[1], ':', row[5]) # print name and number of ratings

Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie Talkie : 27662
Miitomo : 2

In [24]:
for row in ios_final:
    if row[-5] == 'Music':
        print(row[1], ':', row[5]) # print name and number of ratings

Pandora - Music & Radio : 1126879
Spotify Music : 878563
Shazam - Discover music, artists, videos & lyrics : 402925
iHeartRadio – Free Music & Radio Stations : 293228
SoundCloud - Music & Audio : 135744
Magic Piano by Smule : 131695
Smule Sing! : 119316
TuneIn Radio - MLB NBA Audiobooks Podcasts Music : 110420
Amazon Music : 106235
SoundHound Song Search & Music Player : 82602
Sonos Controller : 48905
Bandsintown Concerts : 30845
Karaoke - Sing Karaoke, Unlimited Songs! : 28606
My Mixtapez Music : 26286
Sing Karaoke Songs Unlimited with StarMaker : 26227
Ringtones for iPhone & Ringtone Maker : 25403
Musi - Unlimited Music For YouTube : 25193
AutoRap by Smule : 18202
Spinrilla - Mixtapes For Free : 15053
Napster - Top Music & Radio : 14268
edjing Mix:DJ turntable to remix and scratch music : 13580
Free Music - MP3 Streamer & Playlist Manager Pro : 13443
Free Piano app by Yokee : 13016
Google Play Music : 10118
Certified Mixtapes - Hip Hop Albums & Mixtapes : 9975
TIDAL : 7398
YouTube Mu

The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

In [25]:
for row in ios_final:
    if row[-5] == 'Reference':
        print(row[1], ':', row[5]) # print name and number of ratings

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating. 

However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

- Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

- Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

- Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

## On Google Play
In order to find out what genres are the most popular (have the most users), we will look at number of installs for each app genre. For the Google Play data set, we can find this information in the `Installs` column

In [26]:
display_table(android_final, 5)

1,000,000+:15.72653429602888
100,000+:11.552346570397113
10,000,000+:10.548285198555957
10,000+:10.1985559566787
1,000+:8.393501805054152
100+:6.915613718411552
5,000,000+:6.825361010830325
500,000+:5.561823104693141
50,000+:4.772111913357401
5,000+:4.512635379061372
10+:3.542418772563177
500+:3.2490974729241877
50,000,000+:2.3014440433212995
100,000,000+:2.1322202166064983
50+:1.917870036101083
5+:0.7897111913357401
1+:0.5076714801444043
500,000,000+:0.27075812274368233
1,000,000,000+:0.22563176895306858
0+:0.04512635379061372
0:0.01128158844765343


The install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.). For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to find out which app genres attract the most users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on. To perform computations, however, we'll need to convert each install number from a string to a float. This means we need to remove the commas and the plus characters, or the conversion will fail and cause an error.

In [27]:
cat_android = freq_table(android_final, 1)
cat_installs = {}
for cat in cat_android:
    total_installs = 0
    cat_count = 0
    for row in android_final:
        if row[1] == cat:
            installs = row[5]
            installs = installs.replace(',', '')
            installs = installs.replace('+', '')
            installs = float(installs)
            total_installs += installs
            cat_count += 1
    avg_installs = total_installs/cat_count
    cat_installs[cat] = avg_installs
cat_installs = {k:v for k,v in sorted(cat_installs.items(), key= lambda x:x[1], reverse = True)}
for k,v in cat_installs.items():
    print(k, ':', v)
    
        

COMMUNICATION : 38456119.167247385
VIDEO_PLAYERS : 24727872.452830188
SOCIAL : 23253652.127118643
PHOTOGRAPHY : 17840110.40229885
PRODUCTIVITY : 16787331.344927534
GAME : 15588015.603248259
TRAVEL_AND_LOCAL : 13984077.710144928
ENTERTAINMENT : 11640705.88235294
TOOLS : 10801391.298666667
NEWS_AND_MAGAZINES : 9549178.467741935
BOOKS_AND_REFERENCE : 8767811.894736841
SHOPPING : 7036877.311557789
PERSONALIZATION : 5201482.6122448975
WEATHER : 5074486.197183099
HEALTH_AND_FITNESS : 4188821.9853479853
MAPS_AND_NAVIGATION : 4056941.7741935486
FAMILY : 3695641.8198090694
SPORTS : 3638640.1428571427
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1833495.145631068
BUSINESS : 1712290.1474201474
LIFESTYLE : 1437816.2687861272
FINANCE : 1387692.475609756
HOUSE_AND_HOME : 1331540.5616438356
DATING : 854028.8303030303
COMICS : 817657.2727272727
AUTO_AND_VEHICLES : 647317.8170731707
LIBRARIES_AND_DEMO : 638503.734939759
PARENTING : 542603.6206896552
BEAUTY : 51315

In [29]:
for row in android_final:
    if row[1] == 'COMMUNICATION'and (row[5] == '1,000,000,000+'
                                      or row[5] == '500,000,000+'
                                      or row[5] == '100,000,000+'):
        print(row[0],':',row[5]) # name and n_installs

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs.

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:

In [30]:
under_100M = []
for row in android_final:
    n_installs = row[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    n_installs = float(n_installs)
    if (row[1] == 'COMMUNICATION') and (n_installs < 100000000):
        under_100M.append(n_installs)
sum(under_100M)/len(under_100M)

3603485.3884615386

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:

In [31]:
for row in android_final:
    if row[1] == 'BOOKS_AND_REFERENCE':
        print(row[0],':', row[5]) # name and n_installs

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [32]:
for row in android_final:
    if row[1] == 'BOOKS_AND_REFERENCE' and (row[5] == '1,000,000,000+'
                                            or row[5] == '500,000,000+'
                                            or row[5] == '100,000,000+') :
        print(row[0],':', row[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [33]:
for row in android_final:
    n_installs = row[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    n_installs = float(n_installs)
    if (row[1] == 'BOOKS_AND_REFERENCE') and (1000000 <= n_installs < 100000000):
        print(row[0],':', row[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

# Conclusions
In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.